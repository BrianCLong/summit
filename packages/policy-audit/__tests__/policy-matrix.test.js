const fs = require('fs');
const path = require('path');
const { PolicyAudit } = require('..');
const fixtures = require('../../../contracts/policy/fixtures.json');

const policyDir = path.join(__dirname, '../../../contracts/policy');
const auditDir = path.join(__dirname, '../../../tmp-matrix');

jest.setTimeout(60000);

beforeAll(() => {
  fs.rmSync(auditDir, { recursive: true, force: true });
  fs.mkdirSync(auditDir, { recursive: true });
});

afterEach(() => {
  fs.rmSync(auditDir, { recursive: true, force: true });
  fs.mkdirSync(auditDir, { recursive: true });
});

function expand(map, ref) {
  if (!ref) return {};
  if (typeof ref === 'string') {
    return JSON.parse(JSON.stringify(map[ref]));
  }
  return JSON.parse(JSON.stringify(ref));
}

test('policy decision matrix matches snapshot', async () => {
  const pa = new PolicyAudit({ policyDir, auditDir });
  const results = [];
  for (const scenario of fixtures.cases) {
    const subject = expand(fixtures.subjects, scenario.subject);
    const resource = expand(fixtures.resources, scenario.resource);
    const context = expand(fixtures.contexts, scenario.context);
    const decision = await pa.evaluate({
      subject,
      resource,
      action: scenario.action,
      context,
    });
    expect(decision).toEqual(scenario.expect);
    results.push({ name: scenario.name, allow: decision.allow, reason: decision.reason });
  }
  await pa.close();
  expect(results).toMatchSnapshot();
});
