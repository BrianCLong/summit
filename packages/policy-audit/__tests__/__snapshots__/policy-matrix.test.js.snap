// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`policy decision matrix matches snapshot 1`] = `
[
  {
    "allow": true,
    "name": "analyst dossier investigation allow",
    "reason": "authorized",
  },
  {
    "allow": false,
    "name": "analyst dossier disallowed action",
    "reason": "action-not-allowed",
  },
  {
    "allow": false,
    "name": "analyst dossier wrong role",
    "reason": "role-mismatch",
  },
  {
    "allow": false,
    "name": "analyst dossier missing purpose",
    "reason": "invalid-purpose",
  },
  {
    "allow": false,
    "name": "analyst dossier tenant mismatch",
    "reason": "tenant-mismatch",
  },
  {
    "allow": false,
    "name": "analyst ledger license mismatch",
    "reason": "license-mismatch",
  },
  {
    "allow": false,
    "name": "analyst ledger audit attempt",
    "reason": "license-mismatch",
  },
  {
    "allow": false,
    "name": "analyst maintenance tenant mismatch",
    "reason": "tenant-mismatch",
  },
  {
    "allow": false,
    "name": "analyst maintenance wrong tenant context",
    "reason": "tenant-mismatch",
  },
  {
    "allow": false,
    "name": "analyst missing role",
    "reason": "role-mismatch",
  },
  {
    "allow": false,
    "name": "secret analyst insufficient clearance",
    "reason": "insufficient-clearance",
  },
  {
    "allow": false,
    "name": "secret analyst ledger license mismatch",
    "reason": "license-mismatch",
  },
  {
    "allow": false,
    "name": "secret analyst maintenance tenant mismatch",
    "reason": "tenant-mismatch",
  },
  {
    "allow": true,
    "name": "auditor ledger authorized",
    "reason": "authorized",
  },
  {
    "allow": false,
    "name": "auditor ledger wrong purpose",
    "reason": "invalid-purpose",
  },
  {
    "allow": false,
    "name": "auditor ledger wrong role",
    "reason": "role-mismatch",
  },
  {
    "allow": false,
    "name": "auditor ledger tenant mismatch",
    "reason": "tenant-mismatch",
  },
  {
    "allow": false,
    "name": "auditor dossier license mismatch",
    "reason": "license-mismatch",
  },
  {
    "allow": false,
    "name": "auditor maintenance tenant mismatch",
    "reason": "tenant-mismatch",
  },
  {
    "allow": true,
    "name": "contractor maintenance authorized",
    "reason": "authorized",
  },
  {
    "allow": false,
    "name": "contractor maintenance wrong action",
    "reason": "action-not-allowed",
  },
  {
    "allow": false,
    "name": "contractor maintenance wrong purpose",
    "reason": "invalid-purpose",
  },
  {
    "allow": false,
    "name": "contractor dossier clearance",
    "reason": "insufficient-clearance",
  },
  {
    "allow": false,
    "name": "contractor ledger clearance",
    "reason": "insufficient-clearance",
  },
  {
    "allow": false,
    "name": "auditor ledger missing role",
    "reason": "role-mismatch",
  },
]
`;
